Host stable/kube-state-metrics helm chart

scottrigby opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

See #1153 

To-do:
- [x] import stable/kube-state-metrics helm chart history
- [x] clean up chart README
- [x] bump chart version (PATCH is fine, as this chart can upgrade in-place using the new repo info)
- [x] add chart-testing GitHub Action (for chart CI)
- [x] add chart-releaser GitHub Action (for helm repo CD)
    - [ ] configure a gh-pages branch for package releasing
- [ ] invite stable/kube-state-metrics chart maintainers to collaborate on this git repo
    - [ ] add an OWNERS file to match, if needed (I left the one in place from helm/charts. However please provide guidance on what you would like to do with this)

After PR merge:
- ~[ ] follow-up PR: import chart version package history, and update gh-pages index.yaml file (so that users can continue to access older chart package versions after the `stable` repo is deprecated and garbage collected)~ NO LONGER NEEDED now that GitHub has sponsored helm/charts package history. See https://helm.sh/blog/helm-turns-five/ 🎉 
- [ ] add repo to CNCF [Artifact Hub](https://artifacthub.io/)
    - [ ] add Artifact Hub repo metadata file to [verify repo](https://github.com/artifacthub/hub/issues/550) (gh-pages branch)
- ~[ ] [deprecate](https://github.com/helm/charts/blob/master/PROCESSES.md#deprecating-a-chart) stable/kube-state-metrics~ DONE as part of https://github.com/helm/charts/issues/24310
    - ~[ ] close any stable/kube-state-metrics issues, pointing users here ([there are currently none](https://github.com/helm/charts/issues?q=is%3Aissue+is%3Aopen+kube-state-metrics))~ This will be taken care of as part of the helm/charts repo deprecation follow-up.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1153

scottrigby commented:
I believe the CLA problem is because I'm (purposefully) also importing other user's commits from the stable repo.

torstenwalter commented on charts/kube-state-metrics/Chart.yaml:
```diff
@@ -0,0 +1,20 @@
+- monitoring
+- prometheus
+- kubernetes
+version: 2.9.2
```
I suggest to bump version to 2.10.0. That allows to use 2.9.2 in the old repo to deprecate the chart over there.

scottrigby replied:
@nestorsalceda 👋 Good question, I had to think about this for a moment.

Re semver, this chart repo move isn't a backwards incompatible change (MAJOR), or a new feature (MINOR), so PATCH is fine.

In general when charts move repos they don't bump two PATCH versions to address a conflict with the deprecation bump in `stable`. We could do that here, but it give us anything – there is no version conflict since the release will be from a different repo.

If you see a problem with this line of thinking lmk. But I wouldn't think so since this is the way every other chart repo move has done it.

lilic commented on charts/kube-state-metrics/Chart.yaml:
```diff
@@ -0,0 +1,20 @@
+- prometheus
+- kubernetes
+version: 2.9.2
+appVersion: 1.9.7
```
We launched a pre-release candidate v2.0.0-alpha which is now merged in master, if we merge this it will differ.

scottrigby replied:
Hi @lilic. In general it's ok for the appVersion in a chart to be behind the newest release of the app it installs (for a short time anyway, it has to be, since the container image [tag](https://github.com/kubernetes/kube-state-metrics/pull/1237/files#diff-fd2bc408fdc516b05aa2e779a5417fb2R5) for the app must be available for CI to pass for the PR.

But in this case, 1.9.7 is the most recent release, right? https://github.com/kubernetes/kube-state-metrics/releases/tag/v1.9.7

lilic replied:
We have a pre-release of v2.0.0-alpha cut a few days ago, not sure if that counts for helm charts and the meaning of stable in helm? Just for your information https://github.com/kubernetes/kube-state-metrics/releases/tag/v2.0.0-alpha.

torstenwalter commented:
@brancz @lilic @tariq1890 any feedback on this PR.

lilic commented:
@scottrigby sorry I was under the impression is not ready for review yet. I have not used helm recently but will give it a try in reviewing.

scottrigby commented:
@lilic Hi 👋 I'm happy to move this out of draft mode, I was just looking for feedback per https://github.com/kubernetes/kube-state-metrics/issues/1153#issuecomment-695050047 and https://github.com/kubernetes/kube-state-metrics/pull/1237#issuecomment-695049380.

Re CLA, from what I understand CNCF requires either CLA or DCO. Most of these commits have DCO signoff, and for those which don't (since many of these commits were made before DCO was enforced) I could add mine. However I understand Kubernetes contribution guidelines require CLA. I could not hope to get that from all of these contributors, so can we discuss allowable options for retaining this commit history? I believe history is important for several reasons - both credit to those who worked on it as well as a record of past change motivations that may help inform future decisions (especially if we hope to simplify the chart in future due to new options such as Helm's post-renderer flag).

What do you think about me modifying the commits to list myself as author and crediting the original committers as co-authors? I could add a note in the commit message clarifying that modification was made due to CLA requirements, and this way it would still be an accurate depiction of the work, including me collating and brining in the previous work. I would think this would be more acceptable from an attribution point of view than flattening it all and bringing it in as if it were my own commit. If done this way, could my own CLA then suffice to allow retaining commit history?

eddiezane commented:
Maybe @mrbobbytables or @nikhita have suggestions for importing commits from folks that haven't signed the CLA.

nikhita commented:
We've merged one-offs like these (CLA problems due to code migration - https://github.com/kubernetes/kube-openapi/pull/211#issuecomment-722326200) in the past so we could merge it manually. Before we do that, we need to ensure:

- The git history is maintained. This looks good as-is. :white_check_mark: 
- The LICENSE/copyright is handled correctly. I _think_ this probably means that the helm/charts LICENSE needs to be added to the `charts/kube-state-metrics` directory...but IANAL so I'm not sure. :)

@swinslow can you please let us know what would be the right workflow for handling the license/copyright?

fyi @dims

lilic commented:
Let me know if there is anything you need from my side, I have admin access to merge if we get an okay from Kubernetes org.

scottrigby commented:
@nikhita Great news, yes that would be much easier. I have moved this PR from draft to ready for review. I have also updated the PR description checklist now that some of the steps are no longer needed.

eddiezane commented on charts/kube-state-metrics/values.yaml:
```diff
@@ -0,0 +1,161 @@
+# Default values for kube-state-metrics.
+prometheusScrape: true
+image:
+  repository: quay.io/coreos/kube-state-metrics
```
Any chance we can change this to the gcr image and build a 1.9.7 image there?

See https://github.com/prometheus-community/helm-charts/issues/373

scottrigby replied:
@eddiezane Good idea. Though based on past experience I would prefer that to be a follow-up PR to keep the scope clear in this one. Does that sound reasonable?

lilic replied:
1.9.7 did not have the gcr images yet, but you can do the migration in v2.0.0 to gcr.

eddiezane replied:
@scottrigby that works for me.

@lilic Since v2.0.0 will have breaking changes I'd like to advocate for the last v1 release to have an official image (with multi arch support) cut.

scottrigby commented:
I have added one additional commit. This was the final commit made to the stable chart since I opened this PR, and before the deprecation notice: https://github.com/helm/charts/pull/23968. The only difference is now the version is bumped again, in case users have `2.9.3` installed from stable.

See https://github.com/helm/charts/commits/master/stable/kube-state-metrics/Chart.yaml for a list of all changes to this stable chart's Chart.yaml file. I did not add the notice or deprecation PR commits purposefully. This PR is now completely caught up.

scottrigby commented:
@nikhita - wrapping up one more open question:

> * The LICENSE/copyright is handled correctly. I _think_ this probably means that the helm/charts LICENSE needs to be added to the `charts/kube-state-metrics` directory...but IANAL so I'm not sure. :)
> 
> @swinslow can you please let us know what would be the right workflow for handling the license/copyright?

From similar efforts in other k8s projects (example https://github.com/kubernetes/ingress-nginx) the Apache 2.0 License at the root of this repo has been seen as compatible with the same Apache 2.0 License at the root of the now deprecated helm/charts repo.

BUT to be on the safe side, in case it helps move this PR along, I added the one from helm/charts (with "The Helm Authors" copyright line) to the `charts/kube-state-metrics/` directory. I didn't add to the parent `charts/` directory, in case additional charts (for future sub-components, etc) need to be added in the future.

lilic commented:
@scottrigby I can merge manually as soon as we get an okay from someone from the cncf or Kubernetes org that it's legally alright to do so, we had to get this okay in another project we are migrating. Or if you can reach out to the folks who did not sign the CLA to sign it?

nikhita commented:
> Or if you can reach out to the folks who did not sign the CLA to sign it?

It looks like there are at least 17 people who have not signed the CLA. Getting them to sign the CLA wouldn't be feasible...

@swinslow or @caniszczyk can you please take a look and suggest next steps from a legal perspective?

More context in https://github.com/kubernetes/kube-state-metrics/pull/1237#issuecomment-729141312

nikhita commented:
@swinslow @caniszczyk ping on https://github.com/kubernetes/kube-state-metrics/pull/1237#issuecomment-733631466 :)

KlavsKlavsen commented:
Whats left for this to be merged? Are the labels correct about the issues?

lilic commented:
@KlavsKlavsen yes, it appears so, see the latest comment for more info https://github.com/kubernetes/kube-state-metrics/pull/1237#issuecomment-733631466 only thing left is for either lawyers to say we can merge or everyone to sign the CLA.

@scottrigby can you do a force commit to recheck the CLA, just so we are sure of the latest number of people who need to sign the CLA, thanks!

nikhita commented:
> @scottrigby can you do a force commit to recheck the CLA, just so we are sure of the latest number of people who need to sign the CLA, thanks!

We have a prow command for this! :) 

/check-cla

nikhita commented:
> We have a prow command for this! 

Oops please ignore this! I think this doesn't retrigger the cla bot, so yes, a force/empty commit would be the way to re-check. :+1:

lilic commented:
@nikhita  Yes, I remember that existed at some point fairly recently but forgot the actual command, then checked the docs and could not find the command :D Thanks, I thought I was dreaming it was there! 😅

scottrigby commented:
> > We have a prow command for this!
> 
> Oops please ignore this! I think this doesn't retrigger the cla bot, so yes, a force/empty commit would be the way to re-check. 👍

OK, rebased against `master` and force pushed again.

caniszczyk commented:
We would need the existing folks who contributed to sign the CLA, if you can't find them, you could just remove their contribution if it was small, but in generally, if you only have a dozen or so folks to track down that's not bad at all.

nikhita commented:
> if you only have a dozen or so folks to track down that's not bad at all.

@caniszczyk If we are not able to reach them, how long would you suggest waiting for their responses? In the [past](https://github.com/kubernetes/org/issues/2182#issuecomment-740373871), we have had to wait for _months_. Even then, we didn't get responses from everyone. :( 

Would adding a `NOTICE` file with the list of github usernames that haven't signed the CLA be an alternative?

torstenwalter commented:
Many people are waiting for this to be merged as stable/charts repository is deprecated. As long as this does not have a new home no fixes/uodates can be contributed.

kube-state-metrics is often a central piece in monitoring Kubernetes clusters.

lilic commented:
We seem to have 17 people who have not signed the CLA, have you tried reaching out to them @scottrigby to get them to sign it?

I agree that kube-state-metrics needs to be available for folks to have an insight into their Kubernetes clusters, my proposal is to @scottrigby is it possible to have this repo hosted somewhere else outside of the k8s org until the CLA get signed? And we migrate to kube-state-metrics when this happens? Mainly as we already got a bug report for the helm chart, which would be good to fix. https://github.com/kubernetes/kube-state-metrics/issues/1317

scottrigby commented:
OK, from [Slack thread](https://kubernetes.slack.com/archives/C1TU9EB9S/p1607457638053800) with @caniszczyk we have the go ahead as long as a commit documents where the code is coming from 🎉 

> caniszczyk
I'm fine with this as an exception (since it's coming from a CNCF related repo it's lower risk), just ensure you reference where the code came from
>
> scottrigby
> do you mean ok with the incoming commits as-is (to preserve the history), or OK with squashing them all into one commit under my CLA?
>
> caniszczyk
both, really up to you, just document in the commit where the code is coming from

In that case, I'll add this info explicitly to the last commit message, which adds the LICENSE from that repo.

scottrigby commented:
Commit message updated. Back to you @lilic

tariq1890 commented:
> Regarding the OWNERS file, and collaborators, how do you want to handle that? If you're looking for people to help maintain the chart, I'm already an org member and would be happy to add myself to the maintainers list there and in the chart.

I think we can onboard new OWNERS to `kube-state-metrics`, once they have contributed enough through reviews and commits to the whole project. I don't think the size of kube-state-metrics justifies a directory-level OWNERS structure as of now

tariq1890 commented:
I am actually more in favour of binding the release cadence of helm charts to that of the main application itself. But I'll defer to @lilic and @brancz to arrive at a consensus here.

scottrigby commented:
> I am actually more in favour of binding the release cadence of helm charts to that of the main application itself.

SGTM, but to be clear the way chart-releaser works is chart version packages are bound to unique GitHub releases. It doesn't matter if the chart release cadence matches the application or not. What I meant was that by default these releases are named the same as the chart name (which is the same as the kube-state-metrics project name). However, a recent version of chart-releaser allows this to be renamed, so the releases can be obvious as differentiated from the project application itself when they share a git repo like this one. Hope that makes more sense.

scottrigby commented:
> I think we can onboard new OWNERS to kube-state-metrics, once they have contributed enough through reviews and commits and expressed interest to maintain the whole project. I don't think the size of kube-state-metrics justifies a directory-level OWNERS structure as of now

@tariq1890 do you want to wait for other maintainers to weigh in on this too, or should I go ahead and just remove the OWNERS file now?

tariq1890 commented:
@scottrigby Sure, we can definitely wait on their input.

As for the current OWNERs of the helm charts, `fiunchinho` is not an active reviewer of the helm charts and he had actually asked to be removed from the `OWNERs` list.

@mrueg has been a very active contributor and it would be great to have him onboard. If he's okay with it of course.

mrueg commented:
Sure I'd be happy to continue to help out. 

In regards to coupling release cadence, ideally we'd deliver small chart improvements immediately to users. Would we be okay to have several patch releases per month?

Also not sure how helm copes with unstable /prereleases, I assume the average helm user wants to upgrade to a stable version of ksm.

lilic commented:
> I am actually more in favour of binding the release cadence of helm charts to that of the main application itself.

Agreed, but not up to me, I am not a helm user, so I have no clue, but this makes sense to me. I would say let's start with some cadence and we can always change later on?

@scottrigby I was under the impression we wanted to just remove the commits and squash all into one? As we still see a failing CI? e.g. like @mrueg suggested as per this PR https://github.com/kubernetes/ingress-nginx/pull/5159

As for the OWNERS file, I guess its just @tariq1890 and @mrueg in the file then?

scottrigby commented:
> > I am actually more in favour of binding the release cadence of helm charts to that of the main application itself.
> 
> Agreed, but not up to me, I am not a helm user, so I have no clue, but this makes sense to me. I would say let's start with some cadence and we can always change later on?

@lilic I can see that, but generally chart repos don't work like this. With the helm/chart-releaser-action, each time a PR is merged into the main branch, a new version of the chart is released. Charts also normally have a different version than the application (charts generally version faster than the app it installs). If you want to change this workflow to have fewer releases, you could do it by having a staging branch and merge into the main branch less frequently – in that case you would also want to adjust the chart-testing options, otherwise each PR would still need to bump the version, even though there would be fewer actual releases, which would confuse users. In general this would be an anti-pattern for charts. Is it OK for the chart to have its own release cadence, based on need/PR?

> 
> @scottrigby I was under the impression we wanted to just remove the commits and squash all into one? As we still see a failing CI? e.g. like @mrueg suggested as per this PR [kubernetes/ingress-nginx#5159](https://github.com/kubernetes/ingress-nginx/pull/5159)

On this question, @caniszczyk said it's up to us which way we do it: 
- https://kubernetes.slack.com/archives/C1TU9EB9S/p1607542333077700?thread_ts=1607457638.053800&cid=C1TU9EB9S
- https://kubernetes.slack.com/archives/C1TU9EB9S/p1607543625078000?thread_ts=1607457638.053800&cid=C1TU9EB9S

I brought in commits because it has been useful in other projects to have the history. Here is an example (one of many) where someone has asked for commit history to be restored. This is one I happened to help with: https://github.com/goharbor/harbor-helm/issues/48

From what @nikhita said, the fact that CLA bot isn't passing does not have to be a show stopper, as long as we get the go-ahead from @caniszczyk which we now have. I asked @lilic in Slack for an opinion on this (the value of commit history etc), given the above, who said:

> I have no problem one way or another. Sure commit history is important, no clue about how much in helm case here. But if it has proven to be a problem before happy to continue down this path 👍 
> I don’t use helm, so I am the wrong person to ask :smile: I am not a blocker, might be easier to reach out to tariq on this as he is the helm chart maintainer as well as kube-state-metrics maintainer

So @tariq1890 just checking with you on this. Do you have an issue keeping the commit history (for the reasons above), or would you prefer I squash it out?

> 
> As for the OWNERS file, I guess its just @tariq1890 and @mrueg in the file then?

@lilic Sounds good, just pushed that change to remove fiunchinho. Do you want to invite @mrueg to the org?

Also @lilic and @tariq1890 I just pushed a change to update the Helm repo GitHub actions to the most recent versions. I noticed the app releases are signed with a GPG key. As of [v1.0.0](https://github.com/helm/chart-releaser/releases/tag/v1.1.0) `@helm/chart-releaser-action` added the ability to GHG sign releases. If you would like, we could work together on this in a follow up PR?

tariq1890 commented:
@scottrigby I'd say we keep the commit history of the helm charts. It would definitely be unfortunate if we lost a record of all the helm chart contributors and their contributions. Final decision : I am in favour of this.

tariq1890 commented:
Follow up PR for chart signing sounds good to me. I believe this PR should solely be intended for chart migration. CI improvements can be done in subsequent PRs. Will let @lilic chime in as well

lilic commented:
Looks good to me, I am just wondering does the CLA check checks all the code or just the newest commits? So if we merge this will the next PR fail with CLA or will it pass?

Otherwise lgtm 🎉

nikhita commented:
> I am just wondering does the CLA check checks all the code or just the newest commits? So if we merge this will the next PR fail with CLA or will it pass?

It just checks the commits in the PR so future PRs won't have any problems with the CLA. :)

@lilic also please feel free to merge manually once OWNERS is [fixed](https://github.com/kubernetes/kube-state-metrics/pull/1237#issuecomment-744716218)!

lilic commented:
Thanks can do!

@tariq1890 do we want to just leave you in the OWNERS and once @mrueg is part of k8s org, we can add him after that? So we unblock this PR?

mrueg commented:
> Thanks can do!
> 
> @tariq1890 do we want to just leave you in the OWNERS and once @mrueg is part of k8s org, we can add him after that? So we unblock this PR?
Quick update on this: My request to become part of the org was approved and I've accepted the invite. https://github.com/kubernetes/org/issues/2378 Thanks again for sponsoring!

nikhita commented:
/verify-owners

lilic commented:
/lgtm

Thanks all! 🎉

lilic commented:
Merging manually. 🎉 😅

devurandom commented:
The build failed:

> Failed to resolve action download info. Error: Unable to resolve action `helm/chart-releaser-action@v1.1.1`, unable to find version `v1.1.1`

Latest version appears to be v1.1.0: https://github.com/helm/chart-releaser-action/releases

lilic commented:
@devurandom see this comment here, this will be done in a follow PR. https://github.com/kubernetes/kube-state-metrics/pull/1237#issuecomment-745023575 but you are welcome to submit a PR for this!
